src/stanza.c

```diff
@@ -1492,6 +1492,43 @@ xmpp_stanza_t *xmpp_iq_new(xmpp_ctx_t *ctx, const char *type, const char *id)
     return _stanza_new_with_attrs(ctx, "iq", type, id, NULL);
 }
 
+/** Create a `<iq><ping/></iq>` stanza object.
+ *
+ *  @param ctx a Strophe context object
+ *  @param id attribute 'id'
+ *
+ *  @return a new Strophe stanza object
+ *
+ *  @ingroup Stanza
+ */
+xmpp_stanza_t *xmpp_ping_new(xmpp_ctx_t *ctx, const char *id)
+{
+    xmpp_stanza_t *iq = xmpp_iq_new(ctx, "get", id);
+    if (!iq)
+        goto exit;
+
+    xmpp_stanza_t *ping = xmpp_stanza_new(ctx);
+    if (!ping) {
+        xmpp_stanza_release(iq);
+        iq = NULL;
+        goto exit;
+    }
+
+    int ret = xmpp_stanza_set_name(ping, "ping");
+    if (ret == XMPP_EOK)
+        ret = xmpp_stanza_set_ns(ping, XMPP_NS_PING);
+    if (ret == XMPP_EOK)
+        ret = xmpp_stanza_add_child(iq, ping);
+	xmpp_stanza_release(ping);
+    if (ret == XMPP_EOK)
+        goto exit;
+    xmpp_stanza_release(iq);
+    iq = NULL;
+
+exit:
+    return iq;
```

Comment:
How about writing this as follows?

```suggestion
    xmpp_stanza_t *iq = xmpp_iq_new(ctx, "get", id);
    if (!iq)
        return NULL;

    xmpp_stanza_t *ping = xmpp_stanza_new(ctx);
    if (!ping)
        goto error_out;

    if (xmpp_stanza_set_name(ping, "ping") != XMPP_EOK)
        goto error_out;
    if (xmpp_stanza_set_ns(ping, XMPP_NS_PING) != XMPP_EOK)
        goto error_out;
    if (xmpp_stanza_add_child_ex(iq, ping, 0) != XMPP_EOK)
        goto error_out;
    return iq;

error_out:
    if (ping)
	xmpp_stanza_release(ping);
    xmpp_stanza_release(iq);
    return NULL;
```

IMO this looks a lot cleaner and easier to follow.

Reply:
I'm totally fine with any of your decision. I have expected that my suggestion doesn't fit well for libstrophe itself, but I thought that it would be better to share a proposal. A separate library with XEPs is a good idea.